core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1336,8 +1337,15 @@ protected RexLiteral makeLiteral(
       o = ((TimestampWithTimeZoneString) o).round(p);
       break;
     case DECIMAL:
-      if (o != null && type.getScale() != RelDataType.SCALE_NOT_SPECIFIED) {
-        assert o instanceof BigDecimal;
+      if (o == null) {
```

Comment:
is this a problem with literals only, or does it happen to non-constant expressions too?
Can you check if there is a Quidem test (in an .iq file) for this?
If not, can you please add some?

Reply:
For non-constant expressions actual casting performed by `CastImplementor` (see `RexToLixTranslator#checkExpressionPadTruncate`, there is multiplication to time unit multiplier too). But if cast contains trivial expressions it can be simplified to just literal, without call of `CastImplementor`.